lib/Utils/Signal.cpp

```diff
@@ -0,0 +1,21 @@
+//------------------------------------------------------------------------------
+// CLING - the C++ LLVM-based InterpreterG :)
+// author:  Saagar Jha <saagar@saagarjha.com>
+//
+// This file is dual-licensed: you can choose to license it under the University
+// of Illinois Open Source License or the GNU Lesser General Public License. See
+// LICENSE.TXT for details.
+//------------------------------------------------------------------------------
+
+#include "cling/Utils/Signal.h"
+
+#include "llvm/Support/Signals.h"
+
+jmp_buf before_execution;
+
+void InterruptHandler() {
+  // LLVM automatically resets the handler after every call. Set it back to our
```

Comment:
Is there no way for us to teach llvm not to do that? If not, can we re-use their handler?

Reply:
I'm not completely sure, but from what I saw there didn't seem like a way to do this. The LLVM signal handling code, [here](https://github.com/llvm-mirror/llvm/blob/3c8af5480a208e84dc6ffc24285611c15b51177b/lib/Support/Unix/Signals.inc#L228), seems to unconditionally unset the signal handler just before running it:

```c++
void (*IF)() = InterruptFunction;
Guard.unlock();
InterruptFunction = nullptr;
IF();        // run the interrupt function.
```